include/drivers/rtio.h

```
@@ -0,0 +1,482 @@
+};
+
+/**
+ * @brief Allocate a block from a mempool
+ */
+static inline int rtio_mempool_block_alloc(
+	struct rtio_block_allocator *allocator, struct rtio_block **block,
+	size_t size, u32_t timeout)
+{
+	struct rtio_mempool_block_allocator *mempool_allocator =
+		(struct rtio_mempool_block_allocator *)allocator;
+	struct k_mem_block memblock;
+	size_t block_size = size + sizeof(struct rtio_mempool_block);
+
+	int res = k_mem_pool_alloc(mempool_allocator->mempool,
+				   &memblock, block_size,
+				   timeout);
+	if (res == 0) {
+		struct rtio_mempool_block *mempool_block =
+			(struct rtio_mempool_block *)(memblock.data);
+		u8_t *dataptr = (u8_t *)memblock.data + sizeof(*mempool_block);
+
+		mempool_block->id = memblock.id;
+		*block = (struct rtio_block *)mempool_block;
+		rtio_block_init(*block, dataptr, size);
+	}
+	return res;
+}
+
+static inline void rtio_mempool_block_free(
+	struct rtio_block_allocator *allocator,
+	struct rtio_block *block)
+{
+	struct rtio_mempool_block *mempool_block =
+		(struct rtio_mempool_block *)block;
+
+	k_mem_pool_free_id(&mempool_block->id);
+}
+
+/**
+ * @brief Define an rtio block mempool allocator
+ */
+#define RTIO_MEMPOOL_ALLOCATOR_DEFINE(_name, _minsz, _maxsz, _nmax)	\
+	K_MEM_POOL_DEFINE(rtio_pool_mempool_##_name,			\
+			  sizeof(struct rtio_mempool_block) + _minsz,	\
+			  sizeof(struct rtio_mempool_block) + _maxsz,	\
+			  _nmax,					\
+			  4);						\
+									\
+	struct rtio_mempool_block_allocator _name = {			\
+	.allocator = {							\
+	.alloc = rtio_mempool_block_alloc,				\
+	.free = rtio_mempool_block_free					\
+},									\
+	.mempool = &rtio_pool_mempool_##_name				\
+}
+
+
+/**
+ * @brief Allocate a rtio_block from an rtio_pool with a timeout
+ *
+ * This call is not safe to do with a timeout other than K_NO_WAIT
+ * in an interrupt handler.
+ *
+ * The allocator *must* have a memory layout equivalent to
+ * struct rtio_block_allocator
+ */
+static inline int rtio_block_alloc(void *allocatorv,
```

The `void *` will hide all the errors related to invalid pointer type passed.
Why you are not using `struct rtio_block_allocator *` here?

---

To avoid casting in the case where you've used a specific allocator. I could use a macro here instead if thats more sensible

---

I've fixed this by doing some pointer casting in the macro which defines the mem pool and then always passing in a allocator pointer here now